src/gui/painting/qpainter.cpp

```diff
@@ -2673,6 +2673,152 @@ QPainterPath QPainter::clipPath() const
     }
 }
 
+static QPainterPath noClipping(const QTransform &matrix, const QPainterClipInfo &info)
+{
+    return QPainterPath();
+}
+
+static QPainterPath PathClipReplaceClip(const QTransform &matrix, const QPainterClipInfo &info)
+{
+    return info.path * matrix;
+}
+
+static QPainterPath PathClipIntersectClip(const QTransform &matrix, const QPainterClipInfo &info)
+{
+    return PathClipReplaceClip(matrix, info);
+}
+
+static QPainterPath RectClipReplaceClip(const QTransform &matrix, const QPainterClipInfo &info)
+{
+    QPainterPath tempPath;
+    tempPath.addRect(info.rect);
+    tempPath = tempPath * matrix;
+    return tempPath;
+}
+
+static QPainterPath RectClipIntersectClip(const QTransform &matrix, const QPainterClipInfo &info)
+{
+    return RectClipReplaceClip(matrix, info);
+}
+
+static QPainterPath RectFClipReplaceClip(const QTransform &matrix, const QPainterClipInfo &info)
+{
+    QPainterPath tempPath;
+    tempPath.addRect(info.rectf);
+    tempPath = tempPath * matrix;
+    return tempPath;
+}
+
+static QPainterPath RectFClipIntersectClip(const QTransform &matrix, const QPainterClipInfo &info)
+{
+    // Use rect intersection if possible.
+    QPainterPath tempPath;
+
+    if (matrix.type() <= QTransform::TxScale) {
+        tempPath.addRect(matrix.mapRect(info.rectf));
+    } else {
+        tempPath.addRegion(matrix.map(QRegion(info.rectf.toRect())));
+    }
+    return tempPath;
+}
+
+static QPainterPath RegionClipReplaceClip(const QTransform &matrix, const QPainterClipInfo &info)
+{
+    QPainterPath tempPath;
+    tempPath.addRegion(info.region * matrix);
+    return tempPath;
+}
+
+static QPainterPath RegionClipIntersectClip(const QTransform &matrix, const QPainterClipInfo &info)
+{
+    return RegionClipReplaceClip(matrix, info);
+}
+
+enum ActionForClipOperation { 
+    NoClip, 
+    ReplaceClip, 
+    IntersectClip 
+};
+
+typedef QPainterPath (*PainterFunc)(const QTransform &matrix, const QPainterClipInfo &info);
+
+PainterFunc clipped_path[][IntersectClip + 1] = {
+    { noClipping, RegionClipReplaceClip, RegionClipIntersectClip },
+    { noClipping, PathClipReplaceClip, PathClipIntersectClip },
+    { noClipping, RectClipReplaceClip, RectClipIntersectClip },
+    { noClipping, RectFClipReplaceClip, RectFClipIntersectClip }
+};
```

Comment:
I feel this might be clearer by first creating the clipped_path array with size `[RectFClip + 1][IntersectClip + 1]` then setting the individual rows to their values here (not setting every individual element). Otherwise it's a little unclear that the rows are indexed by clip type. I got that from having read it a lot already, someone in the future might not.

I'd also probably change the function names to have lower case first letters, to match other functions, and change clipped_path to `clippedPath` (or actually more like `clipFuncs` which is clearer to me what it's doing). Not a huge issue though.

Overally this is looking really good though, much tidier and easier to understand.